src/tests/algorithms_test/calorimetry_CalorimeterHitDigi.cc

```diff
@@ -44,7 +44,7 @@ TEST_CASE( "the clustering algorithm runs", "[CalorimeterHitDigi]" ) {
     cfg.pedMeanADC = 123;
     cfg.resolutionTDC = 1.0 * dd4hep::ns;
     algo.applyConfig(cfg);
-    algo.init(detector.get(), logger);
+    algo.init(logger);
```

Comment:
Here, the detector instance is trivial, but how would we go about providing it in other cases?

Reply:
It's all being handled through the algorihtm, which is what specifies which services it needs. So, e.g. CalorimeterIslandCluster would get a class member `const algorithms::GeoSvc& m_geoSvc = algorithms::GeoSvc::instance()` which would then expose `m_geoSvc.detector()`. We can even cache that as `m_detector` and access everything the way we do now.

The new bit here is the ServiceSvc_service which connects DD4hep_service to algorithms::ServiceSvc so it knows that algorithms::GeoSvc is to be provided by DD4hep_service.

Reply:
In a sense it is allowing the algorithm to express which services it needs, and allowing EICrecon to specify which services it can provide. This is better than having the interface of algorithms (i.e. `init()`) depend on having factories pass resources from the services.

Reply:
But this is not EICrecon, this is a unit test.

Reply:
Ah, ok, I see what you are asking. Let me figure that one out and modify the test to use the (trivial) detector. We will need to provide a (trivial) ServiceSvc ourselves, I think.

Reply:
That's what concerns me a bit. If we need to rely on how symbols are resolved (to full ServiceSvc or mock one), that might be a bit fragile.

Reply:
Fortunately the algorithms service infrastructure is much more lightweight than JANA's and doesn't depend on symbol resolution in the same way (or at all, depending on how specifically you are defining symbol resolution).

It's all based on a singleton pattern so in a single test process it is entirely independent of other services and it only relies on C++ language features, not dlopen. See 20aa4ca.